apiserver/facades/agent/secretsmanager/secrets.go

```diff
@@ -63,12 +64,41 @@ type SecretsManagerAPIV1 struct {
 	*SecretsManagerAPI
 }
 
-func (s *SecretsManagerAPI) canRead(uri *coresecrets.URI, entity names.Tag) (bool, error) {
-	return commonsecrets.CanRead(s.secretsConsumer, s.authTag, uri, entity)
+func accessorFromTag(accessorTag names.Tag) secretservice.SecretAccessor {
+	var accessor secretservice.SecretAccessor
+	accessorName := accessorTag.Id()
+	switch kind := accessorTag.Kind(); kind {
+	case names.UnitTagKind:
+		accessor.UnitName = &accessorName
+	case names.ApplicationTagKind:
+		accessor.ApplicationName = &accessorName
+	}
+	return accessor
 }
 
-func (s *SecretsManagerAPI) canManage(uri *coresecrets.URI) (leadership.Token, error) {
-	return commonsecrets.CanManage(s.secretsConsumer, s.leadershipChecker, s.authTag, uri)
+func (s *SecretsManagerAPI) canRead(ctx context.Context, uri *coresecrets.URI, entity names.Tag) (bool, error) {
+	// First try looking up unit access.
+	hasRole, err := s.secretsConsumer.GetSecretAccess(ctx, uri, accessorFromTag(entity))
+	if err != nil {
+		// Typically not found error.
+		return false, errors.Trace(err)
+	}
+	if hasRole.Allowed(coresecrets.RoleView) {
+		return true, nil
+	}
+
+	// All units can read secrets owned by application.
+	appName := commonsecrets.AuthTagApp(s.authTag)
+	hasRole, err = s.secretsConsumer.GetSecretAccess(ctx, uri, secretservice.SecretAccessor{ApplicationName: &appName})
+	if err != nil {
+		// Typically not found error.
+		return false, errors.Trace(err)
+	}
+	return hasRole.Allowed(coresecrets.RoleView), nil
+}
+
+func (s *SecretsManagerAPI) canManage(ctx context.Context, uri *coresecrets.URI) (leadership.Token, error) {
+	return commonsecrets.CanManage(ctx, s.secretsConsumer, s.leadershipChecker, s.authTag, uri)
```

Comment:
These secret access stuff should be in the service as well.

Reply:
I'll add a todo - the PR was getting too large